public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx

```diff
@@ -0,0 +1,526 @@
+      store.setObject(LAST_USED_LABELS_KEY, selectedLabels);
+      if (selected) {
+        // Refetch values for newly selected label...
+        if (!label.values) {
+          this.fetchValues(name);
+        }
+      } else {
+        // Only need to facet when deselecting labels
+        this.doFacetting();
+      }
+    });
+  };
+
+  onClickValue = (name: string, value: string | undefined, event: React.MouseEvent<HTMLElement>) => {
+    const label = this.state.labels.find(l => l.name === name);
+    if (!label || !label.values) {
+      return;
+    }
+    // Toggling value for selected label, leaving other values intact
+    const values = label.values.map(v => ({ ...v, selected: v.name === value ? !v.selected : v.selected }));
+    this.updateLabelState(name, { values }, () => this.doFacetting(name));
+  };
+
+  updateLabelState(name: string, updatedFields: Partial<SelectableLabel>, cb?: () => void) {
+    this.setState(state => {
+      const labels: SelectableLabel[] = state.labels.map(label => {
+        if (label.name === name) {
+          return { ...label, ...updatedFields };
+        }
+        return label;
+      });
+      return { labels };
+    }, cb);
+  }
+
+  componentDidMount() {
+    const { languageProvider } = this.props;
+    if (languageProvider) {
+      const selectedLabels: string[] = store.getObject(LAST_USED_LABELS_KEY, []);
+      languageProvider.start().then(() => {
+        const labels: SelectableLabel[] = languageProvider
+          .getLabelKeys()
+          .slice(0, MAX_LABEL_COUNT)
+          .map(label => ({ name: label, selected: selectedLabels.includes(label), loading: false }));
+        this.setState({ labels }, () => {
+          this.state.labels.forEach(label => {
+            if (label.selected) {
+              this.fetchValues(label.name);
+            }
+          });
+        });
+      });
+    }
+  }
+
+  doFacetting = (lastFacetted?: string) => {
+    const selector = buildSelector(this.state.labels);
+    if (selector === EMPTY_SELECTOR) {
+      // Clear up facetting
+      const labels: SelectableLabel[] = this.state.labels.map(label => {
+        return { ...label, facets: 0, values: undefined, hidden: false };
+      });
+      this.setState({ labels }, () => {
+        // Get fresh set of values
+        this.state.labels.forEach(label => label.selected && this.fetchValues(label.name));
+      });
+    } else {
+      // Do facetting
+      this.fetchSeries(selector, lastFacetted);
+    }
+  };
+
+  async fetchValues(name: string) {
+    const { languageProvider } = this.props;
+    this.updateLabelState(name, { loading: true });
+    try {
+      const values: FacettableValue[] = (await languageProvider.getLabelValues(name))
+        .slice(0, MAX_VALUE_COUNT)
+        .map(value => ({ name: value }));
+      this.updateLabelState(name, { values, loading: false });
+    } catch (error) {
+      console.error(error);
```

Comment:
Is error here likely?  maybe we could use the app events to show an error alert to the user.

Reply:
+1